core/src/main/java/org/apache/druid/data/input/impl/TimedShutoffInputSourceReader.java

```diff
@@ -0,0 +1,141 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input.impl;
+
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.data.input.InputRowListPlusRawValues;
+import org.apache.druid.data.input.InputSourceReader;
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.common.io.Closer;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.common.parsers.CloseableIterator;
+import org.joda.time.DateTime;
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class TimedShutoffInputSourceReader implements InputSourceReader
+{
+  private static final Logger LOG = new Logger(TimedShutoffInputSourceReader.class);
+
+  private final InputSourceReader delegate;
+  private final DateTime shutoffTime;
+
+  public TimedShutoffInputSourceReader(InputSourceReader delegate, DateTime shutoffTime)
+  {
+    this.delegate = delegate;
+    this.shutoffTime = shutoffTime;
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    final ScheduledExecutorService shutdownExec = Execs.scheduledSingleThreaded("timed-shutoff-reader-%d");
+    final CloseableIterator<InputRow> delegateIterator = delegate.read();
+    return decorateShutdownTimeout(shutdownExec, delegateIterator);
+  }
+
+  @Override
+  public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
+  {
+    final ScheduledExecutorService shutdownExec = Execs.scheduledSingleThreaded("timed-shutoff-reader-%d");
+    final CloseableIterator<InputRowListPlusRawValues> delegateIterator = delegate.sample();
+    return decorateShutdownTimeout(shutdownExec, delegateIterator);
+  }
+
+  private <T> CloseableIterator<T> decorateShutdownTimeout(
+      ScheduledExecutorService exec,
+      CloseableIterator<T> delegateIterator
+  )
+  {
+    final Closer closer = Closer.create();
+    closer.register(delegateIterator);
+    closer.register(exec::shutdownNow);
+    final CloseableIterator<T> wrappingIterator = new CloseableIterator<T>()
+    {
+      /**
+       * Indicates this iterator has been closed or not.
+       * Volatile since there is a happens-before relationship between {@link #hasNext()} and {@link #close()}.
+       */
+      volatile boolean closed;
+      /**
+       * Indicates {@link #next} is valid or not.
+       * Not volatile since {@link #hasNext()} and {@link #next()} are supposed to be called by the same thread.
+       */
+      boolean validNext;
+      /**
+       * Caching the next item.
+       * Not volatile since {@link #hasNext()} and {@link #next()} are supposed to be called by the same thread.
+       */
+      T next;
```

Comment:
I think this could be slightly simplified you just check if `next` is `null` instead of `validNext`?

Reply:
Good point. `next` is never null here. Tidied up.